### PR TITLE
add HasLifecycleHooks for CanUpdateState

### PR DESCRIPTION
**Pull request by noxoua**
- [x]  Changes have been thoroughly tested to not break existing functionality.
- [x]  New functionality has been documented or existing documentation has been updated to reflect changes.

___

Added `HasLifecycleHooks` for `CanUpdateState`.

This allows you to add hooks `before` and `after` for the following columns:

- CheckboxColumn
- SelectColumn
- TextInputColumn
- ToggleColumn

Example:

```php
TextInputColumn::make()
    ->beforeStateUpdated(function ($record, $state) {
        // Runs before the state is saved to the database.
    })
    ->afterStateUpdated(function ($record, $state) {
        // Runs after the state is saved to the database.
    })
```

### CODE REVIEW

File: packages/tables/src/Columns/Concerns/HasLifecycleHooks.php
Comment (danharrin) [inline packages/tables/src/Columns/Concerns/HasLifecycleHooks.php:1]: Please rename these to `afterStateUpdated()`, `callAfterStateUpdated()` etc

Also, please add them directly to the CanUpdateState trait, we don't need a new trait if we're only using them in one place

File: packages/tables/docs/03-columns/08-text-input.md [hunk 1]
```diff
@@ -35,3 +35,17 @@ use Filament\Tables\Columns\TextInputColumn;
 
 TextInputColumn::make('background_color')->type('color')
 ```
+
+## Lifecycle hooks
```
Comment (danharrin) [inline packages/tables/docs/03-columns/08-text-input.md:39]: Please also add the relevant documentation for other editable columns

Comment (danharrin) [reply packages/tables/docs/03-columns/08-text-input.md:39]: Would also be useful to see how the `$record` can be injected. And the `$state`.

File: packages/tables/src/Columns/Concerns/CanUpdateState.php [hunk 2]
```diff
@@ -30,6 +32,8 @@ public function updateState(mixed $state): mixed
             ]);
         }
```
Comment (danharrin) [inline packages/tables/src/Columns/Concerns/CanUpdateState.php:55]: IMO, the hooks should still be run, even if you're using `updateStateUsing()`

File: packages/tables/src/Columns/Concerns/CanUpdateState.php [hunk 3]
```diff
@@ -30,6 +32,8 @@ public function updateState(mixed $state): mixed
...
 
+        $this->callBefore();
```
Comment (danharrin) [inline packages/tables/src/Columns/Concerns/CanUpdateState.php]: It's probably worth passing in `$this->evaluate(..., ['state' => $state])`, also for after()

File: packages/tables/src/Columns/Concerns/HasLifecycleHooks.php
Comment (noxoua) [reply packages/tables/src/Columns/Concerns/HasLifecycleHooks.php:1]: done ✅

File: packages/tables/src/Columns/Concerns/CanUpdateState.php (hunk 2 above)
Comment (noxoua) [reply packages/tables/src/Columns/Concerns/CanUpdateState.php:55]: done ✅

File: packages/tables/src/Columns/Concerns/CanUpdateState.php (hunk 3 above)
Comment (noxoua) [reply packages/tables/src/Columns/Concerns/CanUpdateState.php]: done ✅

**Pull request merged**